Apps/WatchServer/src/main/java/rapture/server/WatchRunner.java

```diff
@@ -0,0 +1,146 @@
+import java.nio.file.FileSystems;
+import java.nio.file.Paths;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.log4j.Logger;
+
+import rapture.common.RaptureURI;
+import rapture.common.Scheme;
+import rapture.kernel.ContextFactory;
+import rapture.kernel.Kernel;
+
+public class WatchRunner implements Runnable {
+
+    private volatile Thread thread;
+    private WatchService watchService = null;
+    private String dirPath;
+    private static Logger log = Logger.getLogger(WatchRunner.class);
+    private List<Map> config = new ArrayList<Map>();
+
+    public WatchRunner(String path, List<Map> config) {
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            this.dirPath = path;
+            this.config = config;
+        } catch (IOException e) {
+            log.error(e.getMessage() + ":" + e.getCause());
+        }
+    }
+
+    /**
+     * Start a worker thread to listen for directory changes.
+     */
+    public void startThread() {
+        thread = new Thread(this);
+        // TODO: support each watch to have a poll frequency
+        try {
+            Paths.get(dirPath).register(watchService, ENTRY_CREATE, ENTRY_MODIFY, ENTRY_DELETE);
+            log.info("Monitoring directory: " + dirPath);
+        } catch (IOException e) {
+            log.error(e.getMessage() + ":" + e.getCause());
+        }
+        thread.start();
+    }
+
+    /**
+     * Flag worker thread to stop gracefully.
+     */
+    public void stopThread() {
+        if (thread != null) {
+            Thread runningThread = thread;
+            thread = null;
+            runningThread.interrupt();
+        }
+    }
+
+    @Override
+    public void run() {
+        while (true) {
+            WatchKey watchKey = null;
+            try {
+                watchKey = watchService.take();
+                if (watchKey != null) {
+                    for (WatchEvent<?> watchEvent : watchKey.pollEvents()) {
+                        if (watchEvent.kind() == OVERFLOW) {
+                            log.error("Overflow event count: " + watchEvent.count());
+                            continue;
+                        }
+                        // TODO: Add support to run scripts with an event.
+                        // TODO: Check filepath is correct
+                        String actionUri = getActionForEvent(this.dirPath, watchEvent.kind().name());
+                        if (!actionUri.isEmpty()) {
+                            String filepath = this.dirPath + watchEvent.context();
+                            String scheme = new RaptureURI(actionUri).getScheme().name();
+                            Map<String, String> params = new HashMap<String, String>();
+                            params.put("filetoupload", filepath);
```

Comment:
Just FYI - You can do this in one line as
Map<String, String> params = ImmutableMap.of("filetoupload", this.dirPath + watchEvent.context());